### ISSUE TITLE
Create unique metrics for prometheus metrics that use labels

### ISSUE THREAD

**Issue by garethahealy** [hawkular/hawkular-openshift-agent#105]
As per mail @ http://lists.jboss.org/pipermail/hawkular-dev/2016-December/003533.html

Certain prometheus metrics use labels to distinguish between each datapoint, as if they are unique metrics. Currently, HOSA creates 1 metric and tagged datapoints which when viewed through grafana is displayed as if it were 1 metric.

**Comment by jmazzitelli** [comment]
I'm been thinking about this for the past 2 days - its not trivial to refactor the code to support this :) I'll keep at and maybe I'll get it. The major problems are getting the metric definitions defined and getting the IDs properly expanded (keeping in mind declared IDs in configmaps can have tokens such as ${METRIC:name} or ${POD:name})

**Comment by jmazzitelli** [comment]
I'm doing some final testing on PR #117  but so far so good - I think that PR will provide this feature.

**Comment by jmazzitelli** [comment]
@garethahealy see if you can try out this PR #117. I've done some testing over here and it seems to be working.

**Comment by garethahealy** [comment]
@jmazzitelli will give #117 a try this week (hopefully tonight). Only minor thing i noticed from looking through the change was there was no update to the readme docs.

**Comment by jmazzitelli** [comment]
> i noticed from looking through the change was there was no update to the readme docs.

@garethahealy ah! right. Forgot. I'll take your changes from the other PR and use that. thanks.

**Comment by jmazzitelli** [comment]
@garethahealy  @pilhuhn  - from a discussion with @pilhuhn :

```
(08:59:19 AM) pilhuhn: why not sorting the labels alphabetically by key, and then going for basename_label1key_label1val_label2_key ...
(09:01:09 AM) pilhuhn: I think we should allow to do that, as it saves the user a lot of typing.
```

So the idea is, why force the user to even define the ID with ${label} at all? Instead, the default ID will be as @pilhuhn  suggests: sort the labels alphabetically by key, and then set the ID to "basename_label1key_label1val_label2_key_label2_value_..." (I need to see if h-metrics has a limit on metric ID size)

This does two things:

1) in the case where the user wants to collect ALL metrics, they just don't put in a metrics: section (saves typing, as @pilhuhn  says, and it means the agent will collect any new metrics that might be introduced by the pod later without requiring the admin to know and update the config manually himself). But in this case (not specifying a metrics: section) there is no way for the user to enter his own "id:" field thus no way of specifying how to split out metrics into different time series metric definitions.

2) it follows the default behavior of prometheus itself. Prometheus by definition creates a separate time series for each combination of labels. We should do this by default, too (since the endpoint emitting the metrics is going to assume that anyway - that's how they are designing the metrics they are emitting - one time series per label combination).

We can still support the user defining his own ID and directly putting in his own ${label}s (since this will allow the user to change even the basename if they want - perhaps they want a more descriptive metric name than that given by the metric endpoint itself - this gives them a way to do that).

I'm going to work on that now unless there is opposition.

**Comment by pilhuhn** [comment]
On 31 Jan 2017, at 15:31, John Mazzitelli wrote:

> "basename_label1key_label1val_label2_key_label2_value_..." (I need to 
> see if h-metrics has a limit on metric ID size)

Thinking more about it, should we to make querying easier
different separators  like e.g. `basename_key1=val1,key2=val2`

**Comment by jmazzitelli** [comment]
or what about `basename{key=val1,key1=val2}`

**Comment by jmazzitelli** [comment]
I have it working:

`
I0131 14:28:00.089475   13108 metrics_collector_manager.go:235] TRACE: Metric [test_method_invocations_total] to be split into separate metrics using ID [test_method_invocations_total{method_name=${method_name}}] for endpoint [X|X|prometheus|http://127.0.0.1:8181]
`

At this point, the ID will be as if the user explicitly defined the ID as `test_method_invocations_total{method_name=${method_name}}`

**Comment by jmazzitelli** [comment]
OK, the PR #117 is ready to merge. I just want to talk about how this works. I updated the README with info, too.

First, if you do not specify ANY metrics in the Prometheus endpoints definition, the agent will collect all of them. This actually provides the solution to issue #113.

Second, if a prometheus endpoint emits a metric that has labels (see gareth's example he added to the README.adoc), by default the agent will split that metric into several metrics (each unique combination of prometheus labels becomes a separate metric/metric definition). You do not have to do anything for this to occur (no need to define the id: field in the yaml). If, however, you want control over what those new metric definition IDs look like, you can define your own id: field and put in the ${x} tokens where x is a prometheus label key. For example:

```
metrics:
 - name: jvm_memory_pool_bytes_committed
   id: jvm_memory_pool_bytes_committed_${pool}
```

In this example, if you did not define this "id" field, the default will be:

` jvm_memory_pool_bytes_committed{pool=${pool}}'

**Comment by garethahealy** [comment]
I gave this PR (didn't include your latest naming: {pool=${pool}) a once over last night and it looked good. Will try the latest latest today :)

**Comment by garethahealy** [comment]
Am happy with this PR.

**Comment by jmazzitelli** [comment]
This feature is now in master branch. It is included in the last release 1.1.0.Final found on docker hub.
